src/wrappers.jl

```diff
@@ -1,32 +1,80 @@
 
-#TODO: composed preconditioners, preconditioner setter for cache,
-#   detailed tests for wrappers
-
 ## Preconditioners
 
-struct ScaleVector{T}
-    s::T
-    isleft::Bool
+default_preconditioner(s, isleft) = isleft ? I * s : I * (1/s)
+
+"""
+ C  * x = P  * Q  * x
+ Ci * x = Qi * Pi * x
+"""
+struct ComposePreconditioner{Ti,To}
+    inner::Ti
+    outer::To
 end
 
-function LinearAlgebra.ldiv!(v::ScaleVector, x)
+Base.eltype(A::ComposePreconditioner) = Float64 #eltype(A.inner)
+Base.adjoint(A::ComposePreconditioner) = ComposePreconditioner(A.outer', A.inner')
+Base.inv(A::ComposePreconditioner) = InvComposePreconditioner(A)
+
+function LinearAlgebra.mul!(y, A::ComposePreconditioner, x)
+    @unpack inner, outer = A
+    tmp = similar(y)
```

Comment:
don't allocate in the `mul!` implementation. How is this avoided in the other one?